### PR TITLE
Add an option to SignalR to guarantee ordering of send events

### LINKED ISSUE #9901: Blazor (server-side) System.InvalidOperationException: Received a notification for a rendered batch when not expecting it.
I`ve got an Blazor server-side app deployed as self-contained (dotnet 3.0 preview4) on Azure AppServices.
It seems to run fine on simple Pages (counter with additional buttons or the weather forecast displayed in an pageable datagrid), but it dies after 0-5 UI interactions on a more advanced page. 
In the browser-console the following error occurs and after that error the signalR connection is disconnected (without the automatic reconnection kicking in):
grafik

On my local dev maschine no such error occured even once.

To see this error happening open this [link to the deployed webApp](http://miniaturegolf.azurewebsites.net/MiniatureGolf) and click on some button or dropdowns, after some UI interactions the error should occur in the browser console.

If needed the source code for this app is located [here](https://github.com/marcelgoldstein/MiniatureGolf).

Same error as reported in this comments:
[https://devblogs.microsoft.com/aspnet/asp-net-core-updates-in-net-core-3-0-preview-4/#comments-367](https://devblogs.microsoft.com/aspnet/asp-net-core-updates-in-net-core-3-0-preview-4/#comments-367)

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes https://github.com/aspnet/AspNetCore/issues/9901

### CODE REVIEW

File: src/SignalR/clients/ts/signalr/src/HubConnection.ts
Comment (pranavkm) [inline src/SignalR/clients/ts/signalr/src/HubConnection.ts]: Null refs

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts
Comment (pranavkm) [inline src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: Are there other places that needs to be fixed?

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts
Comment (pranavkm) [inline src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: How should this work if the send fails?

File: src/Components/Browser.JS/src/Boot.Server.ts
Comment (halter73) [inline src/Components/Browser.JS/src/Boot.Server.ts]: Are we making this non-default because of perf considerations? If so, that could probably be mitigated by batching multiple sends per post requests if we start falling behind.

I like making this the default, because it provides more consistency between the WebSocket transport vs. SSE/LP.

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts
Comment (halter73) [reply src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: I think we should continue dequeuing after send failures. The error should be reported back to the caller though.

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts
Comment (halter73) [reply src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: I don't think so.

Comment (halter73) [reply src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: I wonder if it wouldn't be better to put the data itself in a queue and create an async send loop that dequeues the data. If there's no data available, the loop could start waiting on a promise that gets tripped on the next send.

The benefit of this is that batching sends into a single post should be made much easier. I talked to @davidfowl about this, and I think we can just concatenate/append all the data and send it in a single call to transport.send which should result in a single post.

File: src/SignalR/clients/ts/signalr/src/HubConnection.ts
Comment (halter73) [reply src/SignalR/clients/ts/signalr/src/HubConnection.ts]: Good catch. Is there a regression test?

Comment (pranavkm) [reply src/SignalR/clients/ts/signalr/src/HubConnection.ts]: I haven't written any tests as yet. I'll get a few done once I had the design out of the way

File: src/Components/Browser.JS/src/Boot.Server.ts
Comment (pranavkm) [reply src/Components/Browser.JS/src/Boot.Server.ts]: @BrennanConroy suggested making this opt-in due to perf considerations. If we do want to make this default:

* Is the value "true" in the options by default
* Or is there an inverse option and the user has a gesture to turn it off?  `disableHttpTransportSendQueue: true`?

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts
Comment (pranavkm) [reply src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: That seems reasonable. What do I do to send a batch of data - send an array?

**Comment by jkotalik** [comment]
Should blazor.js be there? I can't review this PR at all...

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts
Comment (anurse) [reply src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: Hrm. A send loop seems tricky to do right in the single-threaded JavaScript environment

File: src/Components/Browser.JS/package.json
Comment (rynowak) [inline src/Components/Browser.JS/package.json]: should `@aspnet/signalr` still be here?

File: src/Components/test/testassets/ComponentsApp.App/Shared/NavMenu.razor
Comment (rynowak) [inline src/Components/test/testassets/ComponentsApp.App/Shared/NavMenu.razor]: ???

File: src/Components/Browser.JS/src/Boot.Server.ts
Comment (halter73) [reply src/Components/Browser.JS/src/Boot.Server.ts]: Ideally, we get the perf good enough with this change that we don't need an option to disable it. If that doesn't work, I think opt-out is better than opt-in.

Comment (BrennanConroy) [reply src/Components/Browser.JS/src/Boot.Server.ts]: I don't remember complaining about perf here.

The main reason for opt-in is that it's a change in behavior, although it's a change in behavior to be more correct.
I'm ok with it being on by default. cc @anurse

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts
Comment (BrennanConroy) [inline src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: Now that this is a delegate you can't safely assume `this.transport` is not null

File: src/Components/Browser.JS/package.json
Comment (pranavkm) [reply src/Components/Browser.JS/package.json]: We get it transitively from signalr-protocol-msgpack

File: src/Components/test/testassets/ComponentsApp.App/Shared/NavMenu.razor
Comment (pranavkm) [reply src/Components/test/testassets/ComponentsApp.App/Shared/NavMenu.razor]: The link doesn't exist. When I was trying out my super awesome script to test this: https://github.com/aspnet/AspNetCore/pull/10431#issuecomment-494602536, navigating to this link results in a 404 so I ended up removing it.

File: src/SignalR/clients/ts/signalr-protocol-msgpack/package.json
Comment (pranavkm) [inline src/SignalR/clients/ts/signalr-protocol-msgpack/package.json]: @anurse since you were investigating this.

Comment (anurse) [reply src/SignalR/clients/ts/signalr-protocol-msgpack/package.json]: We can't put a `file:` reference in here as-is. NPM doesn't translate this at all during packing (it ends up in the final package as a `file:` link to a path that doesn't exist). If we do this, we have to rewrite the dependency manually at pack time (in `eng/targets/Npm.targets`)

I couldn't find anything built-in to help us out here :(.

Comment (anurse) [reply src/SignalR/clients/ts/signalr-protocol-msgpack/package.json]: There are tools like [Lerna](https://github.com/lerna/lerna) that some repos use to handle this, but I'm hesitant to bring a whole new tool into play here for this...

**Comment by anurse** [comment]
Preview 6 deadline is looming and we still want to validate this change. @pranavkm do you expect to be able to land this today or early tomorrow?

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts
Comment (halter73) [reply src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: As long as failures get reported to the caller, which it looks like it does, I think it's fine for the send queue to continue on failure.

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts
Comment (pranavkm) [reply src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: We should set up a new promise in `send` if we aren't sending. Nothing really would be using this once this is no longer buffering.

Comment (halter73) [reply src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: I agree. I don't know why you'd clear this.

An a tangentially related note, it might be nicer if it wasn't nullable, and it was initialized to a completed promise.

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts
Comment (halter73) [inline src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: I think this would be easier to follow if this was written as an async loop instead of using `.then()`.

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts
Comment (BrennanConroy) [reply src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: Are `.then` callbacks run inline when a promise finishes?

If no, then there is a race here when a callback has been registered and another send comes in right after a send completes.

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts
Comment (halter73) [inline src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: The next call to sendBufferedData won't be called inline. Effectively it's called with a 0 setTimeout. If there's an intervening call to TransportSendQueue.send() it will see that this.sending is false and allow concurrent/out-of-order data.

The fix for this is usually to unset the flag (if necessary since more data could have been buffered while you were being scheduled) after the continuation awaiting the promise starts. In this case that would be sendBufferedData, but it's possible that there's no queued call to sendBufferedData.

That's one reason writing the async loop might be cleaner.

**Review by pranavkm — COMMENTED** [review]
🆙 📅

File: src/SignalR/clients/ts/signalr-protocol-msgpack/package.json
Comment (pranavkm) [reply src/SignalR/clients/ts/signalr-protocol-msgpack/package.json]: I updated the Npm.targets to support rewriting the file. Unfortunately `ReadAllLines` returns a sequence of trimmed lines, so the formatting for the package.json is lost. Is that something to be concerned about?

File: eng/targets/Npm.Common.targets
Comment (pranavkm) [inline eng/targets/Npm.Common.targets]: cc @natemcmaster

File: src/SignalR/clients/ts/signalr-protocol-msgpack/package.json
Comment (anurse) [reply src/SignalR/clients/ts/signalr-protocol-msgpack/package.json]: I am not worried about that. It's not ideal, but it's OK.

File: eng/targets/Npm.Common.targets
Comment (anurse) [reply eng/targets/Npm.Common.targets]: This is awesome @pranavkm. I'm not a huge fan of how we have to do it, but I'm glad to knock #10307 out as part of this.

For context @natemcmaster : Neither NPM nor Yarn seem to have any mechanism to have a "file" reference (i.e. Project Reference) and then convert it to a version reference during pack. So we have to do it in MSBuild...

There are tools ([Lerna](https://github.com/lerna/lerna)) that can make this better, but they are complicated and I don't think adding new tools is a great idea ;).

File: src/SignalR/clients/ts/signalr-protocol-msgpack/signalr-protocol-msgpack.npmproj
Comment (anurse) [inline src/SignalR/clients/ts/signalr-protocol-msgpack/signalr-protocol-msgpack.npmproj]: We should use an npm version range here (see [npm-semver for syntax](https://docs.npmjs.com/misc/semver)). If we want the same behavior we have in NuGet, we need to explicitly specify `>=`. Alternatively we could use `~` which only allows the patch version to float. I think for now, let's match our NuGet behavior:

```suggestion
    <ReplacePackageContent Include="file:../signalr" ReplaceWith=">=$(PackageVersion)" />
```

**Review by analogrelay — COMMENTED** [review]
Go ahead and mark this PR as fixing #10307 as well! Two birds with one stone!

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts
Comment (BrennanConroy) [inline src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: Does this make a member variable on the class and automatically assign it?

Not sure how I feel about it...

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts
Comment (BrennanConroy) [inline src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: Nice! This does looks a lot cleaner

File: src/SignalR/clients/ts/signalr/tests/HttpConnection.test.ts
Comment (BrennanConroy) [inline src/SignalR/clients/ts/signalr/tests/HttpConnection.test.ts]: nit: comment is slightly wrong

File: src/SignalR/clients/ts/signalr/tests/HttpConnection.test.ts
Comment (BrennanConroy) [inline src/SignalR/clients/ts/signalr/tests/HttpConnection.test.ts]: nit: comment is slightly wrong

File: src/SignalR/clients/ts/signalr/tests/HttpConnection.test.ts
Comment (BrennanConroy) [inline src/SignalR/clients/ts/signalr/tests/HttpConnection.test.ts]: Add a test for trying to queue two different types, `string` and `ArrayBuffer`

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts
Comment (BrennanConroy) [inline src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: We already have this class somewhere

Comment (pranavkm) [reply src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: It's in the test `Utiils`.  I do feel slightly icky about deriving from `Promise` when technically what we need is closer to `SemaphoreSlim.WaitAsync`

**Comment by pranavkm** [comment]
🆙 📅

**Review by BrennanConroy — APPROVED** [review]
Should add a test for `queue.stop()`

File: eng/targets/Npm.Common.targets
Comment (natemcmaster) [reply eng/targets/Npm.Common.targets]: Unfortunately, I don't think this will work on macOS/Linux due to https://github.com/microsoft/msbuild/issues/1024.

Minimal repro of the problem -- run this on Linux. The output has the backslash flipped to a forward slash. `{ "key": "value /" with a quote" }`.
```xml
<Project>
    <Target Name="Build">
        <PropertyGroup>
            <Content>{ "key": "value \" with a quote" }</Content>
        </PropertyGroup>

        <WriteLinesToFile Lines="$(Content)" File="out1.txt" Overwrite="true" />
    </Target>
</Project>
```

Comment (pranavkm) [reply eng/targets/Npm.Common.targets]: Is there a way to write custom tasks with arcade? Additionally, could this wait for the next release? I don't think we publish npm packages built from xplat environments to npmjs, do we?

File: src/Components/Browser.JS/Microsoft.AspNetCore.Components.Browser.JS.npmproj
Comment (SteveSandersonMS) [inline src/Components/Browser.JS/Microsoft.AspNetCore.Components.Browser.JS.npmproj]: I'm sure this is all for a good reason but could you explain what it is?

Previously the `WebpackInputs`/`WebPackOutputs` were for incremental build. Is that handled in another way now, or are we losing that feature?

Is the `ProjectReference` to `signalr-protocol-msgpack.npmproj` just a way of ensuring the SignalR library builds first, or are we changing to a source reference instead of an NPM reference now?

File: src/Components/Browser.JS/package.json
Comment (SteveSandersonMS) [inline src/Components/Browser.JS/package.json]: Would this be problematic if we choose to publish the Blazor package to NPM? When someone references our package, where is NPM going to pull this dependency from?

NPM has other (but less convenient) ways of doing source-to-source references during development that interact properly with package references in production, i.e., `npm link`. I'm not saying that's necessarily what we need to do right now, but let's be sure this is not creating a problem for ourselves.

File: src/Components/Browser.JS/utils/package.json
Comment (SteveSandersonMS) [inline src/Components/Browser.JS/utils/package.json]: I don't know why this is being added. Let's discuss.

File: src/Components/Browser.JS/package.json
Comment (SteveSandersonMS) [reply src/Components/Browser.JS/package.json]: OK, I see from another comment lower down that there's some kind of `package.json`-rewriting trickery going on to deal with this. Having so much customization over the build process is a worrisome trap we keep falling into, but I don't necessarily have a better suggestion in this case.

If you could at least confirm that the rewriting is what avoids this problem that would be good to know!

File: src/Components/Browser.JS/Microsoft.AspNetCore.Components.Browser.JS.npmproj
Comment (pranavkm) [reply src/Components/Browser.JS/Microsoft.AspNetCore.Components.Browser.JS.npmproj]: Sorry, did not see your comments until after I merged. There's a couple of issues here: since we're compiling SignalR source files, we have to also include that closure to determine incrementalism. However these project files currently don't have a way to specify that. Secondly, restoring a `file://` results in the directory being copied to the `node_modules`. Arcade discovers these directories and attempts to build them, which works out poorly since none of the other files \ folders required to build these project files exist. 

Building this is much slower than earlier, is that something we want to try track and address?

File: src/Components/Browser.JS/utils/package.json
Comment (pranavkm) [reply src/Components/Browser.JS/utils/package.json]: Another issue of the `file://` and maybe I'm just going about it wrong. On Windows, a reference with `file://` simply gets copied to the `node_modules` directory. Changing SignalR source files doesn't cause `yarn install` to invalidate or update it. The build currently cleans up the node_modules and runs `yarn install` as part building, this file exists to install `rimraf`

File: src/Components/Browser.JS/Microsoft.AspNetCore.Components.Browser.JS.npmproj
Comment (SteveSandersonMS) [reply src/Components/Browser.JS/Microsoft.AspNetCore.Components.Browser.JS.npmproj]: > Building this is much slower than earlier, is that something we want to try track and address?

Yes. Obviously not for preview 6, but this is worth following up on.

We're taking on a lot of technical debt around NPM. The amount of nonstandardness in our build system, and its tendency to always increase over time, is problematic.

As soon as preview 6 ships and hence SignalR's NPM package is on the public npmjs.org, can we go back to having a regular package dependency on the published package? These two areas of the product should be pretty decoupled and it would help if we could eliminate the `package.json`-rewriting and source-to-source nonstandardness.

File: src/Components/Browser.JS/utils/package.json
Comment (SteveSandersonMS) [reply src/Components/Browser.JS/utils/package.json]: > On Windows, a reference with file:// simply gets copied to the node_modules directory.

In that case it doesn't seem like a valid way of having a dependency at all. An NPM package, once published, isn't necessarily in the same shape as its pre-publish source files. We wouldn't know whether our packages are really going to work when published.

How pressing is our need to do source-to-source references for NPM packages?

 * If this is the only case, we could sidestep the issue entirely (at least for the forseeable future) by waiting for the SignalR package to be published (preview 6) then reverting to a package reference.
 * Longer term, if we can accept non-atomic updates, we could continue using package references but with whatever private NPM registry feed our CI pushes to. Some PRs would have to be two-phase, which I know isn't ideal, but keeps us completely in the standard happy path for NPM/Yarn.
 * If we absolutely have to have atomic repo-wide PRs, we could have the build publish NPM packages (`.tgz` files) to some artifacts location, presumably with a timestamp as the version, and reference them locally from there. It's not standard, but it's closer to standard than `file://`+hacks.

File: eng/targets/Npm.Common.targets
Comment (natemcmaster) [reply eng/targets/Npm.Common.targets]: Arcade doesn't support a 'repotasks' convention. We don't publish npm packages from xplat, but we would like for the repo to build correctly on all platforms.

File: src/Components/Browser.JS/utils/package.json
Comment (anurse) [reply src/Components/Browser.JS/utils/package.json]: > * If we absolutely have to have atomic repo-wide PRs, we could have the build publish NPM packages (`.tgz` files) to some artifacts location, presumably with a timestamp as the version, and reference them locally from there. It's not standard, but it's closer to standard than `file://`+hacks.

That is certainly an ideal way to do it, unfortunately my understanding is that it would take a lot of hacking Arcade to do it. I believe pack phase that generates the `.tgz` happens after the build phase for all the projects completes.

> How pressing is our need to do source-to-source references for NPM packages?

Since you embed the source directly, not doing a source-to-source reference would effectively mean you'd always be at least a build behind (since you'd only be able to use the publish outputs from the previous build at best). If we did the `.tgz` hack above we could get something a little cleaner and a little better, but I'm not sure what it would take to get there.

> An NPM package, once published, isn't necessarily in the same shape as its pre-publish source files.

This is true, but generally only in relation to `npmignore` files or `package.json`'s `files` element. NPM in fact requires that the source effectively be a superset of what is packaged, by the time `npm pack` is run. The only thing you can really do is filter **out** certain files via an exclude list (`npmignore`) or an include list ([`files`](https://docs.npmjs.com/files/package.json#files)).

Comment (SteveSandersonMS) [reply src/Components/Browser.JS/utils/package.json]: > Since you embed the source directly, not doing a source-to-source reference would effectively mean you'd always be at least a build behind

Yep, that's what I meant with "non-atomic updates" (sorry for being obscure).

While not perfect, I think this is a better situation to be in than fighting against both NPM/Yarn and Arcade and having weird hacks like extra `package.json` roots to work around restore issues.

Perhaps longer term we could request the functionality we need in Arcade to support multiple NPM packages properly (dropping `.tgz` artifacts as each one is built, so they can take package dependencies on each other). Or we go all-in on lerna or a similar alternative that addresses this in a sane way.

Comment (anurse) [reply src/Components/Browser.JS/utils/package.json]: > While not perfect, I think this is a better situation to be in

I'm not sure I agree. How would you get a coherent final build in this scenario? You'd be shipping 3.0 RTM while embedding the previous build of SignalR. In order to fix a Blazor-blocking bug in SignalR you'd either have to do two build cycles (somehow disabling Blazor in the first cycle) or manually flow the dependency (which feels hackier than what we're doing here). Not to mention the fact that we have extra lag introduced in even detecting possibly-breaking bugs. Blazor test failures could be due to a change introduced in an earlier build, causing confusion when investigating them.

It seems really odd to me to be shipping as a single product unit in the same repo and yet have dependent versions drifting between components within that same repo.

Also, it's not at all uncommon to have to use extra tools to manage this stuff. Angular's packages are [full of placeholder versions](https://github.com/angular/angular/blob/815443313044e0ec0ec87af588d09eda6b88f0c8/packages/platform-server/package.json#L15) that get replaced. They do things a little differently, but the premise is the same.

Comment (SteveSandersonMS) [reply src/Components/Browser.JS/utils/package.json]: OK, you're probably right that the drawbacks of that approach outweigh the benefits. I'll be interested to see how the alternative `link://`-based approach that @pranavkm has changed this to works out.

**Pull request merged**
